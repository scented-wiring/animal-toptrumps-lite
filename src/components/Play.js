import { useEffect, useState } from "react";
import { Link } from "react-router-dom";
import { cards } from "../cards/index";
import Card from "../components/Card";
import GameUI from "../components/GameUI";
import "../styles/Play.css";

const Play = () => {
  const [width, setWidth] = useState(window.innerWidth);

  const [playerCards, setPlayerCards] = useState([]);
  const [playerCard, setPlayerCard] = useState({});
  const [computerCards, setComputerCards] = useState([]);
  const [computerCard, setComputerCard] = useState({});
  const [lostCard, setLostCard] = useState("");
  const [cardHighField, setCardHighField] = useState("");
  const [showWonCard, setShowWonCard] = useState("");

  const [playerTurn, setPlayerTurn] = useState(true);

  const [tieCards, setTieCards] = useState([]);
  const [win, setWin] = useState("");
  const [gameWinner, setGameWinner] = useState("");
  const [noCards, setNoCards] = useState(false);

  const handleWinnerCards = (
    loserCard,
    winnerCards,
    setWinnerCard,
    setWinnerCards,
    playerWin
  ) => {
    playerWin ? setShowWonCard("Player") : setShowWonCard("Computer");
    setLostCard(loserCard.name);
    winnerCards.push(winnerCards.splice(0, 1)[0]);
    winnerCards.push(loserCard);
    setWinnerCards(winnerCards);
    setWinnerCard(winnerCards[0]);
    tieCards.length > 0 && setWinnerCards(winnerCards.concat(tieCards));
    if (playerWin) {
      setPlayerTurn(true);
      setWin("Player");
    } else {
      setPlayerTurn(false);
      setWin("Computer");
    }
  };

  const handleLoserCards = (loserCards, setLoserCards, setLoserCard) => {
    loserCards.splice(0, 1);
    setLoserCards(loserCards);
    setLoserCard(loserCards[0]);
  };

  const findHighValue = () => {
    //below is required to remove "id" key
    let cardValues = Object.assign(
      {},
      {
        cool: computerCard.cool,
        handsome: computerCard.handsome,
        largeness: computerCard.largeness,
      }
    );
    let highValue = Math.max(...Object.values(cardValues));
    setCardHighField(
      Object.keys(computerCard).find((key) => computerCard[key] === highValue)
    );
  };

  const handlePlayField = (field) => {
    if (playerCard[field] === computerCard[field]) {
      //tie
      setTieCards((tieCards) => [...tieCards, playerCard, computerCard]);
      handleLoserCards(computerCards, setComputerCards, setComputerCard);
      handleLoserCards(playerCards, setPlayerCards, setPlayerCard);
      setWin("Tie");
    } else if (playerCard[field] > computerCard[field]) {
      //player win
      handleWinnerCards(
        computerCard,
        playerCards,
        setPlayerCard,
        setPlayerCards,
        true
      );
      handleLoserCards(computerCards, setComputerCards, setComputerCard);
    } else {
      //computer win
      handleWinnerCards(
        playerCard,
        computerCards,
        setComputerCard,
        setComputerCards,
        false
      );
      handleLoserCards(playerCards, setPlayerCards, setPlayerCard);
    }
  };

  const handleClearWin = () => {
    setWin(false);
    setShowWonCard("");
    if (win !== "Tie") {
      setTieCards([]);
    }
    if (playerCards.length === 0) {
      setGameWinner("Computer");
    } else if (computerCards.length === 0) {
      setGameWinner("Player");
    } else {
      findHighValue();
    }
  };

  useEffect(() => {
    const deal = (array) => {
      let i;
      for (i = 0; i < array.length; i++) {
        if (array.indexOf(array[i]) % 2 === 0) {
          playerCards.push(array[i]);
        } else {
          computerCards.push(array[i]);
        }
      }
    };

    const shuffle = (array) => {
      var currentIndex = array.length,
        temporaryValue,
        randomIndex;

      while (0 !== currentIndex) {
        randomIndex = Math.floor(Math.random() * currentIndex);
        currentIndex -= 1;

        temporaryValue = array[currentIndex];
        array[currentIndex] = array[randomIndex];
        array[randomIndex] = temporaryValue;
      }

      return array;
    };

    deal(shuffle(cards));
    setPlayerCard(playerCards[0]);
    setComputerCard(computerCards[0]);
    !playerCards.length && !computerCards.length && setNoCards(true);
  }, []);

  useEffect(() => {
    const handleResize = () => {
      setWidth(window.innerWidth);
    };
    window.addEventListener("resize", handleResize);
    return (_) => {
      window.removeEventListener("resize", handleResize);
    };
  });

  const renderPlayerDeck = () => {
    return (
      <div className="player">
        <div className={playerTurn ? "score-active" : "score"}>
          Player: {playerCards.length}{" "}
          {playerCards.length === 1 ? `card` : `cards`}
        </div>
        {showWonCard === "Player" ? (
          <Card {...playerCards.find((card) => card.name === lostCard)} /> // Displays the card that the player has won
        ) : (
          <Card {...playerCard} deckSize={playerCards.length} />
        )}
      </div>
    );
  };

  const renderComputerDeck = () => {
    return (
      <div className="computer">
        <div className={!playerTurn ? "score-active" : "score"}>
          Computer: {computerCards.length}{" "}
          {computerCards.length === 1 ? `card` : `cards`}
        </div>
        {showWonCard === "Computer" ? (
          <Card {...computerCards.find((card) => card.name === lostCard)} /> // Displays the card that the player has won
        ) : (
          <Card {...computerCard} deckSize={computerCards.length} hide={true} />
        )}
      </div>
    );
  };

  const renderGameUI = () => {
    return (
      <GameUI
        playerTurn={playerTurn}
        playField={handlePlayField}
        win={win}
        lostCard={lostCard}
        clearWin={handleClearWin}
        cardHighField={cardHighField}
        gameWinner={gameWinner}
        tieCards={tieCards}
        tieCardsLength={tieCards.length}
        noCards={noCards}
      />
    );
  };

  const arrangeDisplay = () => {
    if (width > 690) {
      return (
        <div className="game">
          {renderPlayerDeck()}
          {renderGameUI()}
          {renderComputerDeck()}
        </div>
      );
    } else {
      return (
        <div className="game">
          <div className="mobile-cards">
            {renderPlayerDeck()}
            {renderComputerDeck()}
          </div>
          {renderGameUI()}
        </div>
      );
    }
  };

  return (
    <div className="Play">
      {arrangeDisplay()}
      <Link to="/">
        <div className="back">&larr; BACK</div>
      </Link>
    </div>
  );
};

export default Play;
