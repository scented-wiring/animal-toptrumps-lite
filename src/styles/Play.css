.Play {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-around;
  height: 100%;
  width: 100%;
}

.game {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  width: 100%;
}

.player {
  margin: 0 5px 0 20px;
}

.computer {
  margin: 0 20px 0 5px;
}

.score {
  display: flex;
  flex-direction: column;
  font-family: "Staatliches", sans-serif;
  font-size: 30px;
  text-align: center;
  margin-bottom: 10px;
}

.score-active {
  display: flex;
  flex-direction: column;
  font-family: "Staatliches", sans-serif;
  font-size: 30px;
  text-align: center;
  margin-bottom: 10px;
  color: white;
  text-shadow: 3px 3px black;
}

.won-card {
  position: absolute;
}

@media only screen and (max-width: 930px) {
  .score,
  .score-active {
    font-size: 25px;
  }

  .game .Card,
  .game .Card-border,
  .game .no-cards {
    height: 268px;
    width: 175px;
  }

  .game .name {
    font-size: 25px;
    margin-top: 5px;
  }

  .game .aka {
    font-size: 14px;
  }

  .game .alignment {
    font-size: 11px;
    margin-bottom: 10px;
  }

  .game #photo {
    margin: 0 auto 10px auto;
    height: 90px;
    width: 120px;
  }

  .game .stat-name1,
  .game .stat-name2 {
    padding-left: 8px;
    height: 20px;
    font-size: 16px;
  }

  .game .stat1,
  .game .stat2 {
    height: 20px;
    width: 30px;
  }

  .game .footer {
    font-size: 7px;
    margin: 5px 3px 0 3px;
  }

  .game .hidden-card {
    height: 248px;
    width: 155px;
  }

  .game #card-back-text {
    margin: 10px 0 10px 0;
    font-size: 40px;
  }

  .game #card-back-logo {
    height: 50px;
    width: 50px;
  }
}

@media only screen and (max-width: 690px) {
  .game {
    flex-direction: column;
    height: 100%;
  }

  .mobile-cards {
    display: flex;
    justify-content: center;
    height: 100%;
    margin-top: 20px;
  }

  .game button {
    font-size: 15px;
    margin: 0 8px;
    height: 26px;
    width: 100px;
  }
}

@media only screen and (max-width: 430px) {
  .game .Card,
  .game .Card-border,
  .game .no-cards {
    width: 145px;
  }

  .game .aka {
    font-size: 12px;
  }

  .game .alignment {
    font-size: 10px;
  }

  .game .stat-name1,
  .game .stat-name2 {
    height: 17px;
    font-size: 14px;
  }

  .game .stat1,
  .game .stat2 {
    height: 17px;
  }

  .game .footer {
    margin: 7px 3px 0 3px;
  }

  .game .hidden-card {
    width: 125px;
  }

  .score {
    font-size: 20px;
  }

  .score-active {
    font-size: 20px;
  }
}

@media only screen and (max-width: 350px) {
  .game .Card,
  .game .Card-border,
  .game .no-cards {
    width: 110px;
    height: 240px;
  }

  .game .name {
    font-size: 20px;
  }

  .game .aka {
    font-size: 10px;
  }

  .game .alignment {
    font-size: 9px;
  }

  .game #photo {
    margin: 0 auto 5px auto;
    height: 81px;
    width: 108px;
  }

  .game .stat-name1,
  .game .stat-name2 {
    padding-left: 5px;
    height: 16px;
    font-size: 13px;
  }

  .game .stat1,
  .game .stat2 {
    height: 16px;
    width: 25px;
  }

  .game .footer {
    font-size: 7px;
    margin-top: 4px;
  }

  .game .hidden-card {
    width: 90px;
    height: 220px;
  }

  .game #card-back-text {
    font-size: 30px;
  }

  .score {
    font-size: 17px;
  }

  .score-active {
    font-size: 17px;
  }
}
