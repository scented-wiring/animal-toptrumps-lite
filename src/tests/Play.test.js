import Play from "../components/Play";
import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";

describe("Play component", () => {
  test("renders correctly", () => {
    const { asFragment } = render(
      <BrowserRouter>
        <Play />
      </BrowserRouter>
    );
    expect(asFragment).toMatchSnapshot();
  });
});
