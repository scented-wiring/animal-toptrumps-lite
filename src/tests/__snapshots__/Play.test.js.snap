// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Play component renders correctly 1`] = `[Function]`;
